docs/architecture/next/malicious_chunk_producer_and_phase2.md

```diff
@@ -0,0 +1,109 @@
+
+### Challenges, Slashing and Rollbacks
+
+So we're in a pickle - as a malicious chunk was just added to the chain. And Fthat's why need to have mechanisms to automatically recover from such situations: Challenges, Slashing and Rollbacks.
+
+#### Challenge
+
+Challenge is a self-contained proof, that something went wrong in the chunk 
+processing. It must contain all the inputs (with their merkle proof), the code
+that was executed, and the outputs (also with merkle proofs).
+
+Such a challenge allows anyone (even nodes that don't track that shard or have 
+any state) to verify the validity of the challenge.
+
+When anyone notices that a current chain contains a wrong transition - they 
+submit such challenge to the next block producer, which can easily verify it 
+and it to the next block.
+
+Then the validators do the verification themselves, and if successful, they 
+sign the block.
+
+When such block is succesfully signed, the protocol automatically slashes 
+correct nodes (more details below) and initiates the rollback to bring the 
+state back to the state before the bad chunk (so in our case, back to the block 
+produced by `B0`).
+
+
+#### Slashing
+
+Slashing is the process of taking away the part of the stake from validators
+that are considered malicious.
+
+In the example above, we'll definately need to slash the ``C1`` - and potentially also any validators that were tracking that shard and did sign the bad block.
+
+Things that we'll have to figure out in the future:
+* how much do we slash? all of the stake? some part?
+* what happens to the slashed stake? is it burned? does it go to some pool?
+
+#### State rollbacks
+
+// TODO: add
+
+
+## Problems with the current Phase 2 design
+
+### Is slashing painful enough?
+In the example above, we'd succesfully slash the ``C1`` producer - but was it  
+enough?
+
+Currently (with 4 shards) you need around 20k NEAR to become a chunk producer. 
+If we increase the number of shards to 100, it would drop the minimum stake to 
+around 1k NEAR.
+
+In such scenario, by sacrificing 1k NEAR, the malicious node can cause the 
+system to rollback a couple blocks (potentially having bad impact on the bridge 
+contracts etc).
+
+On the other side, you could be a non-malicious chunk producer with a corrupted 
+database (or a nasty bug in the code) - and the effect would be the same - the 
+chunk that you produced would be marked as malicious, and you'd lose your stake 
+(which will be a super-scary even for any legitimate validator).
+
+
+So the open question is - can we do something 'smarter' in the protocol to
+detect the case, where there is 'just a single' malicious (or buggy) chunk 
+producer and avoid the expensive rollback?
+
+(one strawman proposal - if we see that the previous block didn't contain any signatures from validators that were tracking a given shard - mark the block as failed and ignore the chunks from this shard for the next X blocks).
```

Comment:
Irrelevant in the context of this chapter.

Reply:
done.